src/main/java/org/folio/rest/support/CqlQuery.java

```diff
@@ -0,0 +1,28 @@
+package org.folio.rest.support;
+
+import java.util.regex.Pattern;
+
+public final class CqlQuery {
+  private static final Pattern CQL_MATCHES_ALL = Pattern.compile(
+      "^ *id *==? *\\* *$"              // id=*
+      + "|"
+      + "^ *id *==? *\"\\*\" *$"        // id="*"
+      + "|"
+      + "^ *cql.allRecords *= *1 *$");  // cql.allRecords=1
+
+  private final String cql;
+
+  public CqlQuery(String cqlQuery) {
+    cql = cqlQuery;
+  }
+
+  /**
+   * True if {@code cqlQuery} is a CQL expression that is known to always match all records, for
+   * example {@code id==*} or {@code cql.allRecords=1}.
+   * If false is returned the expression may still match all records because this
+   * method covers only a few cases and doesn't consider the existing data.
+   */
+  public boolean isMatchingAll() {
+    return CQL_MATCHES_ALL.matcher(cql).find();
+  }
+}
```

Comment:
I don't like this. Parsing CQL (and SQL) by regular expressions is just a bad idea. I would have preferring letting the CQL2PgJSON parser do the work and if it evaluates to `true` or `TRUE`, it would match. Admitted, that would not work with id= searches.. But that's a very lame way of getting all records anyway.

For example dealing with `(cql.allRecords = ok)` as well.

Reply:
> I don't like this. Parsing CQL (and SQL) by regular expressions is just a bad idea. I would have preferring letting the CQL2PgJSON parser do the work and if it evaluates to `true` or `TRUE`, it would match. 

I agree that would be better.

> Admitted, that would not work with id= searches.. But that's a very lame way of getting all records anyway.

I completely agree with this. I don't think we should try to be clever about figuring out if a query matches all records.

Reply:
The regex pattern matching has been replaced with CQLParser.

If you want the single DELETE_ALL Kafka event been replaced by potentially thousands of single record DELETE Kafka events please create a Jira and I will be handle it in a separate pull request. Replacing DELETE_ALL by multiple DELETEs will be a performance regression that requires approval by the sysops.

Reply:
> Replacing DELETE_ALL by multiple DELETEs will be a performance regression that requires approval by the sysops.

How would this be a regression? As I understand it, the semantically explicit `cql.allRecords` publishes the single message.

Have I misunderstood?

Reply:
You wrote:

> I don't think we should try to be clever about figuring out if a query matches all records.

If we don't figure out whether the query matches all records (for example `cql.allRecords`) we cannot call the fast deleteAllInstances() and always need to call the slow delete(String cql) method. This is a performance regression.

Reply:
> You wrote:
> 
> > I don't think we should try to be clever about figuring out if a query matches all records.
> 
> If we don't figure out whether the query matches all records (for example `cql.allRecords`) we cannot call the fast deleteAllInstances() and always need to call the slow delete(String cql) method. This is a performance regression.

I believe in my other [comment|https://github.com/folio-org/mod-inventory-storage/pull/775#discussion_r886650100] I suggested that I thought that a dedicated message was acceptable for the case where the intent was obvious and explicit (cql.allRecords).

My objection was not to that optimisation when the intention was obvious, it was with the inference of intent with other queries that happened to delete all records.

Reply:
To be very explicit, I think treating cql.allRecords as an explicit way to deliberately delete all records is a good idea and in that case only a specific message could be published.

I don't think that any special provision should be taken for other queries that only happen to delete all records. 

It seems that code is still attempting to do that.

Reply:
I wouldn't consider accommodating only the explicit intent of `cql.allRecords` a regression. This is the replacement for the current API.

I do think the code attempting to infer intent (listed below) about other queries should be removed. I would not consider that a regression, because that functionality did not exist previously.

```
    return "id".equalsIgnoreCase(node.getIndex())
        && StringUtils.equalsAny(base, "=", "==")
        && "*".equals(node.getTerm());
```

Reply:
Since RMB v30.2.3 the section ["CQL: Matching all records" of the RMB documentation](https://github.com/folio-org/raml-module-builder#cql-matching-all-records) declares `id=*` to be officially supported:

> A search matching all records in the target index can be executed with a cql.allRecords=1 (CQL standard) or a id=* (RMB specific) query.

Should we add a restriction to the RMB documentation and the mod-inventory-storage documentation and the Nolana release notes saying that mod-inventory-storage no longer has full support for `id=*` in all cases?

Reply:
> Should we add a restriction to the RMB documentation and the mod-inventory-storage documentation and the Nolana release notes saying that mod-inventory-storage no longer has full support for id=* in all cases?

I think @adamdickmeiss summed that up pretty well:

> Admitted, that would not work with id= searches.. But that's a very lame way of getting all records anyway.

I think it's poor guidance to recommend that folks use a query like `id=*` for matching all records, especially when there is a specific provision for an intent revealing clause in the standard.

I ambivalent about whether mod-inventory-storage allows that CQL or not. My concern here is with implementing specific additional behaviour based upon it. I already think that the special message is a compromise.

I don't think this decision needs to be made for all cases, rather the case that we are exploring here, which is a specific behaviour for deletion only.

There are lots of CQL queries that folks can write that don't make much sense. That is one of the major trade offs with using an open query language rather than defined criteria.

If you want to include documentation to describe how folks should avoid using `id=*` etc for matching all records and for deletion specifically, I would support that endeavour.